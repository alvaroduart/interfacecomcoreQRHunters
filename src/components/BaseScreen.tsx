import React, { ReactNode } from 'react';
import { View, StyleSheet, ViewStyle, StatusBar, DimensionValue, TouchableOpacity } from 'react-native';
import { SafeAreaView } from 'react-native-safe-area-context';
import { Ionicons } from '@expo/vector-icons';
import CurvedBackground from '../components/CurvedBackground';
import theme from '../theme/theme';
import { useNavigation, StackActions } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import { RootStackParamList } from '../navigation/types';

interface BaseScreenProps {
  children: ReactNode;
  backgroundColor?: string;
  showCurvedBackground?: boolean;
  showBackButton?: boolean;
  onBackPress?: () => void;
  curvedBackgroundProps?: {
    color?: string;
    height?: DimensionValue;
    borderRadius?: number;
    position?: 'top' | 'bottom';
  };
  style?: ViewStyle;
}

// Componente base para telas — fornece header/back e fundo curvo configurável
const BaseScreen: React.FC<BaseScreenProps> = ({
  children,
  backgroundColor = theme.colors.background,
  showCurvedBackground = true,
  showBackButton = false,
  onBackPress,
  curvedBackgroundProps = {
    color: theme.colors.primary,
    height: '40%',
    borderRadius: theme.borderRadius.extraLarge,
    position: 'bottom'
  },
  style = {},
}) => {
  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();
  
  const handleBackPress = () => {
    if (onBackPress) {
      onBackPress();
    } else {
      navigation.goBack();
    }
  };
  
  return (
    <SafeAreaView style={[styles.container, { backgroundColor }, style]}>
      <StatusBar translucent backgroundColor="transparent" />
      
      {showBackButton && (
        <TouchableOpacity style={styles.backButton} onPress={handleBackPress}>
          <Ionicons name="arrow-back" size={24} color={theme.colors.text.primary} />
        </TouchableOpacity>
      )}
      
      <View style={styles.contentContainer}>
        {children}
      </View>
      
      {showCurvedBackground && (
        <CurvedBackground
          color={curvedBackgroundProps.color}
          height={curvedBackgroundProps.height}
          borderRadius={curvedBackgroundProps.borderRadius}
          position={curvedBackgroundProps.position}
        />
      )}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  contentContainer: {
    flex: 1,
    padding: theme.spacing.lg,
    zIndex: 2,
  },
  backButton: {
    position: 'absolute',
    top: 40,
    left: 20,
    zIndex: 10,
    padding: 8,
    borderRadius: 20,
    backgroundColor: 'rgba(255, 255, 255, 0.7)',
  },
});

export default BaseScreen;
