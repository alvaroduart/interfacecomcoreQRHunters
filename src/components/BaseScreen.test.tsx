
import React from 'react';
import { render } from '@testing-library/react-native';
import { Text } from 'react-native';
import BaseScreen from './BaseScreen';
import { NavigationContainer } from '@react-navigation/native';

describe('BaseScreen', () => {
  function renderWithNav(children: React.ReactNode) {
    return render(<NavigationContainer>{children}</NavigationContainer>);
  }

  it('renders children correctly', () => {
    const { getByText } = renderWithNav(
      <BaseScreen>
        <Text>Test Child</Text>
      </BaseScreen>
    );
    expect(getByText('Test Child')).toBeTruthy();
  });

  it('applies style and backgroundColor props', () => {
    const { toJSON } = renderWithNav(
      <BaseScreen style={{ backgroundColor: 'red' }}>
        <Text>Styled</Text>
      </BaseScreen>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('shows back button and calls onBackPress', () => {
    const onBackPress = jest.fn();
    const { UNSAFE_queryAllByType } = renderWithNav(
      <BaseScreen showBackButton onBackPress={onBackPress}>
        <Text>Back Test</Text>
      </BaseScreen>
    );
    // Find TouchableOpacity and simulate press
    const touchables = UNSAFE_queryAllByType(require('react-native').TouchableOpacity);
    expect(touchables.length).toBeGreaterThan(0);
    touchables[0].props.onPress();
    expect(onBackPress).toHaveBeenCalled();
  });

  it('renders CurvedBackground when showCurvedBackground is true', () => {
    const { UNSAFE_queryAllByType } = renderWithNav(
      <BaseScreen showCurvedBackground>
        <Text>Curved</Text>
      </BaseScreen>
    );
    expect(UNSAFE_queryAllByType(require('./CurvedBackground').default).length).toBeGreaterThan(0);
  });

  it('does not render CurvedBackground when showCurvedBackground is false', () => {
    const { UNSAFE_queryAllByType } = renderWithNav(
      <BaseScreen showCurvedBackground={false}>
        <Text>No Curve</Text>
      </BaseScreen>
    );
    expect(UNSAFE_queryAllByType(require('./CurvedBackground').default).length).toBe(0);
  });
});
