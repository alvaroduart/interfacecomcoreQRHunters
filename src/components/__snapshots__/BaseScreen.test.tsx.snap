// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseScreen applies style and backgroundColor props 1`] = `
<RCTSafeAreaView
  style={
    [
      {
        "flex": 1,
      },
      {
        "backgroundColor": "#f7f6ed",
      },
      {
        "backgroundColor": "red",
      },
    ]
  }
>
  <View
    style={
      {
        "flex": 1,
        "padding": 24,
        "zIndex": 2,
      }
    }
  >
    <Text>
      Styled
    </Text>
  </View>
  <View
    style={
      [
        {
          "left": 0,
          "position": "absolute",
          "right": 0,
          "zIndex": 1,
        },
        {
          "backgroundColor": "#4169e1",
          "borderTopLeftRadius": 130,
          "borderTopRightRadius": 130,
          "bottom": 0,
          "height": "40%",
        },
        {},
      ]
    }
  />
</RCTSafeAreaView>
`;
